src/MongoDB.Driver.Core/Core/Compression/Zstandard/ZstandardNativeWrapper.cs

```diff
@@ -0,0 +1,286 @@
+        public int RecommendedOutputSize => (int)_recommendedZstreamOutputSize;
+
+        // public methods
+        public void Compress(
+            OperationContext operationContext,
+            int inputCompressedSize,
+            int inputUncompressedSize,
+            out int compressedBytesProcessed,
+            out int uncompressedBytesProcessed)
+        {
+            Ensure.IsNotNull(operationContext, nameof(operationContext));
+            Ensure.That(_compressionMode == CompressionMode.Compress, nameof(_compressionMode));
+
+            InitializeIfNotAlreadyInitialized();
+
+            // compressed data
+            var outputNativeBuffer = CreateNativeBuffer(
+                operationContext.CompressedPinnedBufferWalker, // operation result
+                (ulong)inputCompressedSize);
+
+            // uncompressed data
+            var inputNativeBuffer = CreateNativeBuffer(
+                operationContext.UncompressedPinnedBufferWalker,
+                (ulong)inputUncompressedSize);
+
+            // compress inputNativeBuffer to outputNativeBuffer
+            Zstandard64NativeMethods.ZSTD_compressStream(
+                _zstreamPointer,
+                outputNativeBuffer,
+                inputNativeBuffer);
+
+            compressedBytesProcessed = (int)outputNativeBuffer.Position; // because start Position is always 0
+            uncompressedBytesProcessed = (int)inputNativeBuffer.Position; // because start Position is always 0
+
+            operationContext.UncompressedPinnedBufferWalker.Offset += uncompressedBytesProcessed;
+            // CompressedPinnedBufferWalker.Offset is always 0
+        }
+
+        public void Decompress(
+            OperationContext operationContext,
+            int compressedOffset,
+            int inputCompressedSize,
+            int inputUncompressedSize,
+            out int compressedBytesProcessed,
+            out int uncompressedBytesProcessed)
+        {
+            Ensure.IsNotNull(operationContext, nameof(operationContext));
+            Ensure.That(_compressionMode == CompressionMode.Decompress, nameof(_compressionMode));
+
+            InitializeIfNotAlreadyInitialized();
+
+            // apply reading progress on CompressedPinnedBufferWalker
+            operationContext.CompressedPinnedBufferWalker.Offset = compressedOffset;
+            // compressed data
+            var inputNativeBuffer = CreateNativeBuffer(
+                inputCompressedSize <= 0 ? null : operationContext.CompressedPinnedBufferWalker,
+                inputCompressedSize <= 0 ? 0 : (ulong)inputCompressedSize);
+
+            // uncompressed data
+            var outputNativeBuffer = CreateNativeBuffer(
+                operationContext.UncompressedPinnedBufferWalker, // operation result
+                (ulong)inputUncompressedSize);
+
+            // decompress inputNativeBuffer to outputNativeBuffer
+            Zstandard64NativeMethods.ZSTD_decompressStream(
```

Comment:
`ZSTD_decompressStream` returns a value that we are ignoring.

Reply:
Answered above

Reply:
See further comment above.